server/impl/user_facebook.go

```diff
@@ -0,0 +1,145 @@
+	if err != nil {
+		logger.WithError(err).Error("Failed to exchange for access token from Facebook")
+		return redirectToHome
+	} else if !token.Valid() {
+		logger.WithField("token", token).Error("Facebook provided invalid access token")
+		return redirectToHome
+	}
+
+	logger.WithFields(log.Fields{
+		"token_type": token.TokenType,
+		"access":     token.AccessToken,
+		"refresh":    token.RefreshToken,
+		"expiry":     token.Expiry,
+	}).Info("Successfully logged into Facebook")
+
+	// Step 3: Fetch user information using the access token.
+
+	c := oauth2.NewClient(ctx, oauth2.StaticTokenSource(token))
+	resp, err := c.Get("https://graph.facebook.com/v9.0/me?fields=id%2Cname%2Cemail&format=json")
+	if err != nil {
+		logger.WithField("token", token).WithError(err).Error("Unable to get user information from Facebook")
+		return redirectToHome
+	}
+	defer resp.Body.Close()
+
+	jsonDec := json.NewDecoder(resp.Body)
+	var info struct{ ID, Name, Email string }
+	if err := jsonDec.Decode(&info); err != nil {
+		logger.WithField("token", token).WithError(err).Error("Unable to get user information from Facebook")
+		return redirectToHome
+	}
+
+	logger.Info(info)
+
+	// Step 4.1: Match the Facebook user with an existing user, if possible.
+
+	// Look for a user with this Facebook ID.
+	var dbUser db.User
+	if err := i.DB().First(&dbUser, "facebook_id = ?", info.ID).Error; err != nil && err != gorm.ErrRecordNotFound {
+		logger.WithError(err).Error("Unable to lookup user by Facebook ID")
+		return redirectToHome
+	} else if err == nil {
+		session := SessionFromContext(ctx)
+		session.Values = map[interface{}]interface{}{
+			UserID: dbUser.IDString(),
+		}
+		logger.Info("logged in through facebook ID")
+		return SaveSession(param.HTTPRequest, redirectToHome)
+	}
+
+	// Look for a user with this email. If found, add the Facebook ID to their
+	// account.
+	if err := i.DB().First(&dbUser, "email = ?", info.Email).Error; err != nil && err != gorm.ErrRecordNotFound {
+		logger.WithError(err).Error("Unable to lookup user by email")
+		return redirectToHome
+	} else if err == nil {
+		session := SessionFromContext(ctx)
+		session.Values = map[interface{}]interface{}{
+			UserID: dbUser.IDString(),
+		}
+
+		if dbUser.FacebookID == "" {
+			dbUser.FacebookID = info.ID
+			if err := i.DB().Model(&dbUser).Update("FacebookID", dbUser.FacebookID).Error; err != nil {
+				logger.WithError(err).Warn("Unable to update user's Facebook ID")
+			}
+		}
+
+		logger.Info("logged in through email")
+		return SaveSession(param.HTTPRequest, redirectToHome)
+	}
+
+	// Step 4.2: Create a new user for this Facebook user.
+
+	dbUser.Name = info.Name
+	dbUser.FacebookID = info.ID
+	dbUser.Email = info.Email
+	if err := i.DB().Create(&dbUser).Error; err != nil {
+		logger.WithError(err).Error("Unable to create new user")
+		return redirectToHome
+	}
+
+	if dbUser.ID == 0 {
+		logger.Error("Created user but ID is still empty")
+		return redirectToHome
+	}
```

Comment:
Random Tim question: when does this happen? Are we just handling this just in case?